destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala

```diff
@@ -0,0 +1,221 @@
+      val hySchema = SQLServerHygiene.hygienicIdent(Ident(schema))
+      val tempFragment = Fragment.const0(hySchema.forSqlName) ++ fr0"." ++ Fragment.const0(hyName.forSqlName)
+      val uName = hyName.unsafeForSqlName
+      val uSchema = hySchema.unsafeForSqlName
+
+      new TempTable {
+        def ingest(chunk: Chunk[CharSequence]): ConnectionIO[Unit] =
+          insertChunk(log)(tempFragment, columns, chunk)
+
+        def drop: ConnectionIO[Unit] =
+          ifExists(log)(uName, uSchema.some).option flatMap { results =>
+            if (results.exists(_ === 1)) {
+              (fr"DROP TABLE" ++ tempFragment)
+                .updateWithLogHandler(log)
+                .run
+                .void
+            } else {
+              ().pure[ConnectionIO]
+            }
+          }
+
+        def create: ConnectionIO[Unit] =
+          ifExists(log)(uName, uSchema.some).option flatMap { results =>
+            if (results.exists(_ === 0)) {
+              val columnsObj = createColumnSpecs(columns)
+              (fr"CREATE TABLE" ++ tempFragment ++ fr0" " ++ columnsObj)
+                .updateWithLogHandler(log)
+                .run
+                .void
+            } else {
+              ().pure[ConnectionIO]
+            }
+          }
+
+        private def truncate: ConnectionIO[Unit] =
+          ifExists(log)(uName, uSchema.some).option flatMap { results =>
+            if (results.exists(_ === 1)) {
+              (fr"TRUNCATE TABLE" ++ tempFragment)
+                .updateWithLogHandler(log)
+                .run
+                .void
+            } else {
+              ().pure[ConnectionIO]
+            }
+          }
+
+        private def insertInto: ConnectionIO[Unit] =
+          (fr"INSERT INTO" ++
+            tableFragment ++ fr0" " ++
+            fr"SELECT * FROM" ++
+            tempFragment)
+            .updateWithLogHandler(log)
+            .run
+            .void
+
+        private def rename: ConnectionIO[Unit] =
+          (fr0"EXEC SP_RENAME '" ++
+            tempFragment ++ fr0"', '" ++
+            Fragment.const0(SQLServerHygiene.hygienicIdent(Ident(unsafeName)).unsafeForSqlName) ++ fr0"'")
+            .updateWithLogHandler(log)
+            .run
+            .void
+
+        private def filterTempIds(idColumn: Column[_]): ConnectionIO[Unit] = {
+          val mkColumn: String => Fragment = parent =>
+            Fragment.const0(parent) ++ fr0"." ++
+            Fragment.const0(SQLServerHygiene.hygienicIdent(Ident(idColumn.name)).forSqlName)
+
+          val fragment =
+            fr"DELETE target FROM" ++ tableFragment ++
+            fr" target INNER JOIN" ++ tempFragment ++ fr" temp" ++
+            fr"ON" ++ mkColumn("target") ++ fr0"=" ++ mkColumn("temp")
+
+          fragment.updateWithLogHandler(log).run.void
+        }
+
+        def append: ConnectionIO[Unit] = {
+          val mbFilter = idColumn traverse_ { col => filterTempIds(col) }
```

Comment:
I'm not sure this is safe for non-upsert sinks and if it is (i.e. all appends contain new ids), then it shouldn't be necessary.

Reply:
Right, thank you.